### PR TITLE
Allow flask applications that require database to start

### PR DESCRIPTION
**Pull request by weiiwang01**
The existing design of the database module cannot modify the database configuration unless the Flask application is already initiated. However, for certain Flask applications, initiation is impossible without the proper database configuration in place. This presents a deadlock scenario.

With this pull request, we propose to address this problem by amending the database module to enable it to update the Flask application's database configuration even prior to the application's initiation.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def __init__(self, *args: typing.Any) -> None:
         )
         self._observability = Observability(charm=self, charm_state=self._charm_state)
 
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.config_changed, self.on_config_changed)
```
Comment (yhaliaw) [inline src/charm.py]: Why is this changed to a public method?

File: src/databases.py [hunk 2]
```diff
@@ -11,16 +11,13 @@
...
-    FLASK_SERVICE_NAME,
-    FLASK_SUPPORTED_DB_INTERFACES,
-)
-from exceptions import InvalidDatabaseRelationDataError, PebbleNotReadyError
+from constants import FLASK_DATABASE_NAME, FLASK_SUPPORTED_DB_INTERFACES
 
 logger = logging.getLogger(__name__)
 
+if typing.TYPE_CHECKING:
+    from charm import FlaskCharm
```
Comment (yhaliaw) [inline src/databases.py]: Is the if statement needed?
It seems the import is only used for type checking, but in the team's code base we generally do not do this.

**Review by arturo-seijas — DISMISSED** [review]
LGTM

File: src/charm.py (hunk 1 above)
Comment (weiiwang01) [reply src/charm.py]: Well, because it will be invoked in the `database.py` module, so I have to make it public.

File: src/databases.py (hunk 2 above)
Comment (weiiwang01) [reply src/databases.py]: This is necessary to avoid circular dependency.

File: src/charm.py (hunk 1 above)
Comment (mthaddon) [reply src/charm.py]: Could the database module not do a `charm.on.config_changed.emit()` (not sure if that's how it'd reference a charm object, but you get the idea) to avoid this?

**Comment by mthaddon** [comment]
Could you add a description to the PR? It'd be good to have a bit more context on the change.

File: src/charm.py (hunk 1 above)
Comment (mthaddon) [reply src/charm.py]: Ah, yeah, looks like you should be able to do `self._charm.on.config_changed.emit()`

File: src/databases.py (hunk 2 above)
Comment (mthaddon) [reply src/databases.py]: I think we should include a comment explaining this, as otherwise it might not be obvious to someone looking at it in the future.

File: src/charm.py (hunk 1 above)
Comment (weiiwang01) [reply src/charm.py]: That's a great idea, updated, thanks!

**Comment by weiiwang01** [comment]
> Could you add a description to the PR? It'd be good to have a bit more context on the change.

Added, I also merged the proxy pull request to main, so now the change list in the pull request should only include the modification to the database system.

File: src/databases.py (hunk 2 above)
Comment (weiiwang01) [reply src/databases.py]: Yes, we will have a meeting discussing the circular dependency problem later this day, and hope we can remove them.

File: src/databases.py [hunk 3]
```diff
@@ -137,9 +114,7 @@ def get_uris(self) -> typing.Dict[str, str]:
             # https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/mysql_client/v0/schemas/provider.json
             if not all(data.get(key) for key in ("endpoints", "username", "password")):
                 logger.warning("Incorrect relation data from the data provider: %s", data)
-                raise InvalidDatabaseRelationDataError(
-                    f"Incorrect relation data from the data provider: {data}"
-                )
+                continue
```
Comment (amandahla) [inline src/databases.py]: Why the error is not being raised anymore?
If an application needs database to start and there is no correct relation data, what would happen?

File: src/databases.py (hunk 2 above)
Comment (amandahla) [reply src/databases.py]: This looks similar (exactly?) to what happened in Synapse where we changed for passing a Pebble instance. I see that you are doing this because the charm needs to call config_change ->  _restart_flask_application. Maybe it's necessary to add the same abstraction here.

**Comment by jdkandersson** [comment]
Just a note - this is still being discuss with @gregory-schiano and myself - the PR shouldn't be merged in its current state

**Comment by weiiwang01** [comment]
> Just a note - this is still being discuss with @gregory-schiano and myself - the PR shouldn't be merged in its current state

Updated with the new design.

File: src/databases.py (hunk 2 above)
Comment (weiiwang01) [reply src/databases.py]: This has been removed in the latest commit, thanks!

File: src/databases.py (hunk 3 above)
Comment (weiiwang01) [reply src/databases.py]: In that case, the charm will enter block status with the error message `Webserver configuration check failed, please review your charm configuration or database relation`

File: src/flask_app.py [hunk 4]
```diff
@@ -54,3 +68,53 @@ def flask_environment(self) -> dict[str, str]:
...
+            self._charm.app.status = status
+
+    def restart_flask(self) -> None:
+        """Restart or start the flask service if not started with the latest configuration."""
+        container = self._charm.unit.get_container("flask-app")
+        if not container.can_connect():
+            logger.info("pebble client in the Flask container is not ready")
+            return
+        if not self._charm_state.is_secret_storage_ready:
+            logger.info("secret storage is not initialized, defer config-changed")
```
Comment (arturo-seijas) [inline src/flask_app.py]: The log msg seems a bit misleading. The event is not deferred

File: tests/unit/test_charm_state.py [hunk 5]
```diff
@@ -44,9 +44,8 @@ def test_charm_state_flask_config(
     harness.begin()
     harness.update_config(charm_config)
     charm_state = CharmState.from_charm(
```
Comment (arturo-seijas) [inline tests/unit/test_charm_state.py:49]: Wouldn't it be better to provide a mocked charm instead of harness for the charm state tests?

File: src/flask_app.py (hunk 4 above)
Comment (weiiwang01) [reply src/flask_app.py]: Updated, thanks!

File: tests/unit/test_charm_state.py (hunk 5 above)
Comment (weiiwang01) [reply tests/unit/test_charm_state.py:49]: Yes, it could be, but it will just be too much trouble to mock the configuration retrieving so I opt to use the harness.

Comment (arturo-seijas) [reply tests/unit/test_charm_state.py:49]: Doesn't something like the following do the trick?
```
charm = MagicMock(config=charm_config)

Comment (weiiwang01) [reply tests/unit/test_charm_state.py:49]: I haven't thought about that, updated, thanks!

Comment (arturo-seijas) [reply tests/unit/test_charm_state.py:49]: Could you update also the other tests?

File: src/charm.py [hunk 6]
```diff
@@ -29,7 +24,11 @@
 
 
 class FlaskCharm(ops.CharmBase):
-    """Flask Charm service."""
+    """Flask Charm service.
+
+    Args:
```
Comment (jdkandersson) [inline src/charm.py]: Should this be `attrs` instead?

File: src/charm.py [hunk 7]
```diff
@@ -38,22 +37,23 @@ def __init__(self, *args: typing.Any) -> None:
...
             self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
             return
-        self._flask_app = FlaskApp(charm_state=self._charm_state)
         self._webserver = GunicornWebserver(
             charm_state=self._charm_state,
             flask_container=self.unit.get_container(FLASK_CONTAINER_NAME),
-            flask_app=self._flask_app,
         )
+        self._flask_app = FlaskApp(
+            charm=self, charm_state=self._charm_state, webserver=self._webserver
```
Comment (jdkandersson) [inline src/charm.py:52]: Should `FlaskApp` depend on the charm? Also, didn't we discuss that this shouldn't be a class anymore?

File: src/charm.py [hunk 8]
```diff
@@ -38,22 +37,23 @@ def __init__(self, *args: typing.Any) -> None:
...
+        )
+        self._databases = Databases(charm=self, flask_app=self._flask_app)
+        self._charm_state.database_uris = self._databases.get_uris()
```
Comment (jdkandersson) [inline src/charm.py]: This doesn't look right, shouldn't the database_uris be passed as an argument?

File: tests/unit/test_charm_state.py (hunk 5 above)
Comment (weiiwang01) [reply tests/unit/test_charm_state.py:49]: Update other test cases where harness was only used for creating `charm_state`, thanks!

File: src/charm.py (hunk 6 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: src/charm.py (hunk 7 above)
Comment (weiiwang01) [reply src/charm.py:52]: I found out that the `FlaskApp` also need to change the charm status, so just passing the flask container is not sufficient. I think the `FlaskApp` is still a class in the latest design.

File: src/charm.py (hunk 8 above)
Comment (weiiwang01) [reply src/charm.py]: I can't, another circular dependency.

File: src/charm.py (hunk 7 above)
Comment (jdkandersson) [reply src/charm.py:52]: As discussed, the status change should only be done in either the charm class or observers

File: src/charm.py [hunk 9]
```diff
@@ -38,22 +33,22 @@ def __init__(self, *args: typing.Any) -> None:
...
-
         try:
             self._charm_state = CharmState.from_charm(
                 charm=self, secret_storage=self._secret_storage
             )
         except CharmConfigInvalidError as exc:
             self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
             return
-        self._flask_app = FlaskApp(charm_state=self._charm_state)
         self._webserver = GunicornWebserver(
```
Comment (jdkandersson) [inline src/charm.py:47]: This should also not be a class, right?

File: src/charm.py (hunk 8 above)
Comment (jdkandersson) [reply src/charm.py]: I suggest that we decouple retrieving the database configuration from the database observer so we don't have this issue anymore

Comment (jdkandersson) [reply src/charm.py]: @gregory-schiano

File: src/charm.py (hunk 9 above)
Comment (weiiwang01) [reply src/charm.py:47]: Since the `webserver` doesn't cause any trouble, the modification of that class is not in high priority, we can do that later, thanks!

File: src/charm.py (hunk 8 above)
Comment (weiiwang01) [reply src/charm.py]: FYI, `DataRequires` in `lib/charms/data_platform_libs/v0/data_interfaces.py` can't be instantized twice for the same relation. The singleton property is enforce by `ops` framework.

File: src/charm.py (hunk 7 above)
Comment (weiiwang01) [reply src/charm.py:52]: Updated, thanks!

File: src/charm.py (hunk 8 above)
Comment (jdkandersson) [reply src/charm.py]: What if the database requires instance is passed as an argument to both the database observer and the charm state? That way the state can use it to retrieve the integration data and the observer can use it to listen to events

Comment (weiiwang01) [reply src/charm.py]: I think the code fragmentation problem is worth considering.

Comment (jdkandersson) [reply src/charm.py]: Discussed with Gregory and the `get_uris` can be a function on the database module, just shouldn't be attached to the database observer. There can also be a function to initialise the data structure containing the database requirers classes in the dictionary which can be passed as an argument to the database observer and `get_uris` function. This should address the concern around code fragmentation

Comment (weiiwang01) [reply src/charm.py]: updated.

File: src/charm.py [hunk 10]
```diff
@@ -38,21 +33,26 @@ def __init__(self, *args: typing.Any) -> None:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
-        self._databases = Databases(charm=self)
         self._secret_storage = SecretStorage(charm=self)
-
+        self._database_requirers = Databases.make_database_requirers(self)
```
Comment (jdkandersson) [inline src/charm.py]: I'm guessing this is a static method, should this just be top level function in the module?

File: src/charm.py [hunk 11]
```diff
@@ -38,21 +33,26 @@ def __init__(self, *args: typing.Any) -> None:
...
         try:
             self._charm_state = CharmState.from_charm(
-                charm=self, secret_storage=self._secret_storage
+                charm=self,
+                secret_storage=self._secret_storage,
+                database_uris=Databases.get_uris(self._database_requirers),
```
Comment (jdkandersson) [inline src/charm.py]: Similar to previous comment, should this just be a top level function in the module?

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -22,18 +25,24 @@ def test_flask_pebble_layer(harness: Harness) -> None:
     act: start the flask charm and set flask-app container to be ready.
     assert: flask charm should submit the correct flaks pebble layer to pebble.
     """
-    harness.begin_with_initial_hooks()
-    harness.set_can_connect(harness.model.unit.containers["flask-app"], True)
-    harness.framework.reemit()
+    harness.begin()
+    secret_storage = unittest.mock.MagicMock()
+    secret_storage.is_initialized = True
+    secret_storage.get_flask_secret_key.return_value = "0" * 16
```
Comment (jdkandersson) [inline tests/unit/test_charm.py]: Should this be a constant so that it doesn't need to be repeated in the assert?

**Comment by jdkandersson** [comment]
> Just a note - this is still being discuss with @gregory-schiano and myself - the PR shouldn't be merged in its current state

This is getting close now, thanks for the patience and the updates and feedback! We have made important progress with the architecture guidance in the process of reviewing this PR

File: tests/unit/test_charm.py (hunk 12 above)
Comment (weiiwang01) [reply tests/unit/test_charm.py]: Updated, thanks

File: src/charm.py (hunk 11 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks

File: src/charm.py (hunk 10 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks

**Review by jdkandersson — DISMISSED** [review]

File: src/databases.py [hunk 13]
```diff
@@ -32,121 +97,60 @@ class Databases(ops.Object):  # pylint: disable=too-few-public-methods
...
-    def get_uris(self) -> typing.Dict[str, str]:
-        """Compute DatabaseURI and return it.
+        self._charm.unit.status = status
+        if self._charm.unit.is_leader():
+            self._charm.app.status = status
+
+    def _restart_flask(self) -> None:
+        """Restart or start the flask service if not started with the latest configuration."""
+        try:
+            restart_flask(
```
Comment (amandahla) [inline src/databases.py]: Not in this PR but I just saw that restart_flask does not change the unit status if cannot connect to the container or if the secret is not ready. Won't this impact here?

Comment (weiiwang01) [reply src/databases.py]: I think now the `restart_flask` doesn't change the status at all. The status change is now handled by the caller.

**Review by jdkandersson — APPROVED** [review]

**Review by arturo-seijas — APPROVED** [review]
LGTM
